experimental/dds/tree/src/HistoryEditFactory.ts

```diff
@@ -72,11 +72,16 @@ export function revert(changes: readonly ChangeInternal[], before: RevisionView)
 				const { source } = change;
 				const nodesBuilt = builtNodes.get(source);
 				const nodesDetached = detachedNodes.get(source);
-
 				if (nodesBuilt !== undefined) {
+					if (nodesBuilt.length === 0) {
```

Comment:
Because this is returning early, it will leave the array inside of the `nodesBuilt` map; but I'm pretty sure we want to delete it (as we normally would four lines below). Same thing with `nodesDetached` in the next `else if` case below. There's probably a couple different ways to fix this. I wonder if the most elegant might be to remove both of these checks, and have `nodesBuilt` and `nodesDetached` not be set in the first place (in the Build case on line 50) when the sequence is empty?

Reply:
You could also add an additional test for this bug. Build and insert an empty sequence into a trait, the try to do another build using the same detached sequence ID. Confirm that this current code fails (it should complain when trying to do the second build that there is still a detached sequence, even though there shouldn't be because you just inserted it), and then confirm that whatever you change fixes it.

Reply:
Hello Noah! I created an additional test where I passed a build, insert, then a second build (using the same `DetachedSequenceId`) into the `revert()` function.

i.e. `revert([firstBuild, insertBuild, secondBuild], testTree.view)`

However, the function immediately returns `undefined` during the `insertBuild` change, and therefore does not reach the `secondBuild` (where the code should fail).

Although I wasn't able to produce the error, I tried your suggested solution of removing the two existing checks, and added a check to prevent `nodesBuilt` being set in case of an empty source (In the Build case of line 50). However, I didn't add a check in the Detached case, as the `createInvertedDetach()` function already returns `undefined` before getting the chance to update `nodesDetached`. This change also seems to pass all of the test cases :)

Should we proceed with this solution, as it guarantees no "residual" entries in `nodesBuilt`?

Reply:
Oh you're totally right. I forgot that if one change in the edit fails, we just short circuit and stop the entire revert. Sorry - I hadn't quite pulled the whole picture into my head. This is making me think that we don't want to return `undefined` for these cases; we actually want to skip (i.e. `continue`) that change. When we revert an edit that has a Malformed or Invalid change, we return undefined to indicate that the edit can't be performed. But here, we are trying to introduce revert behavior for inserts/detaches that are Valid, but useless; is that different?

Suppose I have an Edit which does:
1. Build an empty sequence []
2. Insert [] under trait T of some node X 
3. Set payload of some other node Y

I apply this edit, but then I hit "undo" and revert it. With the current logic (returning undefined), it will see that change #2 has an empty sequence, which will return undefined and abort the entire undo edit. But what about setting the payload of node Y? Do we still want to do that? I think we do. The Edit was valid to run forwards, so it seems to me that if we can't run it backwards, _even on the tree state just afterwards_, then we're breaking some kind of implicit and desirable undo/redo invariant (admittedly, not one that we've actually defined AFAIK). What do you think?

And @jenn-le, what do you think? You're experienced with reverts and the consequences of these kinds of decisions.

Reply:
Yeah that's a good point, this makes much more sense to me!

I think the original intent was to let everything else behave the same way, but only for the insertion/detach of empty traits, we do nothing. So I think it makes more sense to `continue` to the next change rather than prematurely returning `undefined`.

Reply:
Yes, that's correct.